case_studies/coadds/coadd_decoder.py

```diff
@@ -0,0 +1,326 @@
+        noiseless_uncentered = torch.zeros(len(dithers), max_n_sources, 1, size, size)
+
+        n_sources = int(full_cat.n_sources[0].item())
+        galaxy_params = full_cat["galaxy_params"][0]
+        plocs = full_plocs[0]
+
+        for ii in range(n_sources):
+            offset_x = plocs[ii][1] + self.bp - size / 2
+            offset_y = plocs[ii][0] + self.bp - size / 2
+            offset = torch.tensor([offset_x, offset_y])
+            centered = self.single_decoder.render_galaxy(
+                galaxy_params[ii], psf, size, dithers=dithers
+            )
+            uncentered = self.single_decoder.render_galaxy(galaxy_params[ii], psf, size, offset)
+            uncentered_dithered = self.single_decoder.render_galaxy(
+                galaxy_params[ii], psf, size, offset, dithers
+            )
+            noiseless_centered[:, ii] = centered.reshape(centered.shape[0], 1, size, size)
+            noiseless_uncentered[:, ii] = uncentered.reshape(uncentered.shape[0], 1, size, size)
+            image0 += uncentered
+            image += uncentered_dithered
+        return image, noiseless_centered, noiseless_uncentered, image0
+
+
+class CoaddGalsimBlends(GalsimBlends):
+    """Dataset of coadd galsim blends."""
+
+    def __init__(
+        self,
+        prior: UniformGalsimGalaxiesPrior,
+        decoder: FullCatalogDecoder,
+        background: ConstantBackground,
+        tile_slen: int,
+        max_sources_per_tile: int,
+        num_workers: int,
+        batch_size: int,
+        n_batches: int,
+        fix_validation_set: bool = False,
+        valid_n_batches: Optional[int] = None,
+    ):
+        super().__init__(
+            prior,
+            decoder,
+            background,
+            tile_slen,
+            max_sources_per_tile,
+            num_workers,
+            batch_size,
+            n_batches,
+            fix_validation_set,
+            valid_n_batches,
+        )
+        self.slen = self.decoder.slen
+        self.pixel_scale = self.decoder.single_decoder.pixel_scale
+
+    def _get_images(self, full_cat, dithers):
+        size = self.slen + 2 * self.bp  # check bp gets passed through
+        (
+            noiseless,
+            noiseless_centered,
+            noiseless_uncentered,
+            image0,
+        ) = self.decoder.render_catalog(
+            full_cat=full_cat, psf=self.decoder.single_decoder.psf, dithers=dithers
+        )
+
+        # align single exposures
+        aligned_images = align_single_exposures(
+            image0[:].reshape(size, size), noiseless, size, dithers
+        )
+
+        # calculate weights
+        weight = 1 / torch.tensor(aligned_images)
+
+        # get background and noisy image.
+        background = self.background.sample(
+            aligned_images.reshape(
+                aligned_images.shape[0], 1, aligned_images.shape[1], aligned_images.shape[2]
+            ).shape
+        )
```

Comment:
try to split this statement into multiple statements that each fit in one line (with e.g. temporary variables)